### DISCUSSION TITLE
Speed up budget optimizer

### DISCUSSION THREAD

**Discussion by AlfredoJF** [pymc-labs/pymc-marketing#1192] [Q&A]
Is there a recommended way to speed up the budget optimizer within pymc-marketing?

I know that providing good initial guesses is beneficial, but has anyone else done something in this space besides that?

**Comment by williambdean** [comment]
Hi @AlfredoJF , 

Good question. There is this thread: https://github.com/pymc-labs/pymc-marketing/issues/780. I suspect the use of grad in scipy optimizer will have some improvements. 

The initial guess kwargs should be available in the [`minimize_kwargs`](https://www.pymc-marketing.io/en/stable/api/generated/pymc_marketing.mmm.mmm.MMM.allocate_budget_to_maximize_response.html#pymc_marketing.mmm.mmm.MMM.allocate_budget_to_maximize_response). I believe that was introduced in 0.10.0. That seems to be the [`x0`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html) kwarg

Carlos is currently working on this PR #1151 which will make some changes which might include a speed up. Feel free to comment in the PR if you have thoughts

**Comment by AlfredoJF** [comment]
Thanks @wd60622!

Unfortunately, `x0` is not part of `minimize_kwargs` but still I think it's ok for now.
https://github.com/pymc-labs/pymc-marketing/blob/fa44cc7fbd134c8a3aa302d0bf2687383f706bff/pymc_marketing/mmm/budget_optimizer.py#L193C9-L193C76

What about setting different params for `minimize_kwargs`? I'm curious about setting `ftol`at that level.
https://github.com/pymc-labs/pymc-marketing/blob/fa44cc7fbd134c8a3aa302d0bf2687383f706bff/pymc_marketing/mmm/budget_optimizer.py#L204C1-L207C14

Might these settings improve the processing time without compromising the allocation? Have we tried before?

```python
minimize_kwargs = {
    "method": "SLSQP",
    "options": {"ftol": 1e-6, "maxiter": 500},
}
```
@carlosagostini  Keen to explore this in the meantime grad and PR https://github.com/pymc-labs/pymc-marketing/pull/1151 are merged.

**Reply by williambdean** [reply]
Makes sense. I overlooked that then.

Would you be interested in making a PR @AlfredoJF . I see you raised a similar issue before: https://github.com/pymc-labs/pymc-marketing/issues/876
